File: bindings/ocaml/test/property.ml
(* C8 native property-based tests for the OCaml facade (qcheck).

   Every property is checked against an in-language brute-force Levenshtein
   oracle:
     (a) distance symmetry d(a,b)=d(b,a), identity d(a,a)=0, and threshold
         consistency dThreshold(a,b,k) = (if d<=k then d else sentinel) across
         the Levenshtein/Damerau/true-Damerau variants, plus standard oracle
         agreement. The over-bound sentinel is derived at runtime rather than
         hard-coded, so the test is robust to its representation as an OCaml int.
     (b) a query's result set at distance k equals {t in dict : lev(query,t)<=k}
         over a random libdictenstein DynamicDawg dictionary, with exact
         distances and value round-trips.
     (c) u64 value round-trips, with 0 and boundary patterns pinned.

   qcheck runs from a fixed Random.State seed, so a failing run reproduces. An
   ASCII alphabet keeps byte = scalar for the oracle; scalar-vs-byte counting is
   pinned by the snapshot suite. Distance operands may be empty: the facade
   marshals an empty string as a null pointer, and as of the LLEV-B18 fix the
   native distance entry points accept NULL+0 as the empty string. *)

module Dict = Vinary_tree_libdictenstein
module Lev = Vinary_tree_liblevenshtein

let alphabet = [ 'a'; 'b'; 'c'; 'd' ]

let term_gen = QCheck.Gen.(string_size ~gen:(oneof_list alphabet) (int_range 0 6))

let term = QCheck.make ~print:(Printf.sprintf "%S") term_gen

(* Derive the over-bound sentinel from a known over-threshold pair. *)
let sentinel = Lev.distance_threshold "aaaa" "bbbb" 0

let oracle a b =
  let la = String.length a and lb = String.length b in
  if a = b then 0
  else if la = 0 then lb
  else if lb = 0 then la
  else begin
    let previous = Array.init (lb + 1) (fun j -> j) in
    let current = Array.make (lb + 1) 0 in
    for i = 1 to la do
      current.(0) <- i;
      for j = 1 to lb do
        let cost = if a.[i - 1] = b.[j - 1] then 0 else 1 in
        current.(j) <-
          min (min (previous.(j) + 1) (current.(j - 1) + 1)) (previous.(j - 1) + cost)
      done;
      Array.blit current 0 previous 0 (lb + 1)
    done;
    previous.(lb)
  end

let text_of_match result =
  match result.Lev.term with
  | Lev.Text text -> text
  | Lev.Tokens _ -> failwith "expected a text match"

let run_query entries query k =
  let dictionary = Dict.dynamic_dawg () in
  List.iter (fun (t, v) -> ignore (Dict.put dictionary t v)) entries;
  let automaton = Lev.transducer (Dict.resource dictionary) in
  let cursor = Lev.query automaton query ~maximum_distance:k in
  let matches = List.of_seq (Lev.to_seq cursor) in
  Lev.cursor_close cursor;
  Dict.close dictionary;
  Lev.close_transducer automaton;
  List.map (fun r -> (text_of_match r, (r.Lev.distance, r.Lev.id))) matches

let distance_property name dist thr =
  QCheck.Test.make ~count:400 ~name (QCheck.triple term term (QCheck.int_range 0 3))
    (fun (a, b, k) ->
      dist a b = dist b a
      && dist a a = 0
      &&
      let full = dist a b in
      let bounded = thr a b k in
      if full <= k then bounded = full else bounded = sentinel)

let oracle_property =
  QCheck.Test.make ~count:400 ~name:"standard oracle agreement" (QCheck.pair term term)
    (fun (a, b) -> Lev.distance a b = oracle a b)

let value_gen =
  QCheck.Gen.(oneof_weighted [ (1, return None); (3, map (fun i -> Some (Int64.of_int i)) (int_bound 1_000_000)) ])

let dictionary_arb =
  QCheck.make
    QCheck.Gen.(list_size (int_range 0 8) (pair term_gen value_gen))

let query_property =
  QCheck.Test.make ~count:150 ~name:"query result set equals oracle"
    (QCheck.triple dictionary_arb term (QCheck.int_range 0 3))
    (fun (raw_entries, query, k) ->
      (* Deduplicate terms (later bindings win), mirroring dictionary insertion. *)
      let table = Hashtbl.create 16 in
      List.iter (fun (t, v) -> Hashtbl.replace table t v) raw_entries;
      let entries = Hashtbl.fold (fun t v acc -> (t, v) :: acc) table [] in
      let got = run_query entries query k in
      let expected = List.filter (fun (t, _) -> oracle query t <= k) entries in
      let got_terms = List.sort compare (List.map fst got) in
      let expected_terms = List.sort compare (List.map fst expected) in
      got_terms = expected_terms
      && List.for_all
           (fun (t, (d, id)) -> d = oracle query t && List.assoc t entries = id)
           got)

let () =
  let rand = Random.State.make [| 20260809 |] in
  let check t = QCheck.Test.check_exn ~rand t in
  check (distance_property "levenshtein" Lev.distance Lev.distance_threshold);
  check (distance_property "damerau" Lev.damerau_distance Lev.damerau_distance_threshold);
  check (distance_property "true_damerau" Lev.true_damerau_distance Lev.true_damerau_distance_threshold);
  check oracle_property;
  check query_property;
  (* (c) u64 value round-trip: 0, 1, and boundary bit patterns. *)
  List.iter
    (fun value ->
      match run_query [ ("term", Some value) ] "term" 0 with
      | [ ("term", (0, id)) ] when id = Some value -> ()
      | _ -> failwith (Printf.sprintf "u64 round-trip failed for %Ld" value))
    [ 0L; 1L; Int64.max_int; Int64.min_int; -1L ];
  print_endline "OCaml property tests passed"

File: bindings/swift/Integration/Sources/SwiftBindingIntegration/Property.swift
import Foundation
import Libdictenstein
import Liblevenshtein

// C8 native property-based tests for the Swift facade: a seeded pseudo-PBT loop
// checked against an in-language Levenshtein oracle.
//
//   (a) distance symmetry d(a,b)==d(b,a), identity d(a,a)==0, and threshold
//       consistency dThreshold(a,b,k) == (d<=k ? d : nil) across the
//       Levenshtein/OSA/true-Damerau variants, plus standard oracle agreement;
//   (b) query result set == {t in dict : lev(query,t) <= k} over a random
//       libdictenstein DynamicDAWG dictionary, with exact distances and value
//       round-trips;
//   (c) u64 value round-trips, with 0 and UInt64.max pinned.
//
// A SplitMix64 RNG seeded from a constant makes each run reproducible. Distances
// count Unicode scalar values, and the alphabet's characters are all single
// scalars, so the oracle over `unicodeScalars` matches the native semantics.

/// Deterministic, seedable RNG (SplitMix64) for reproducible property runs.
struct SeededRNG: RandomNumberGenerator {
    private var state: UInt64
    init(seed: UInt64) { state = seed }
    mutating func next() -> UInt64 {
        state = state &+ 0x9E37_79B9_7F4A_7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58_476D_1CE4_E5B9
        z = (z ^ (z >> 27)) &* 0x94D0_49BB_1331_11EB
        return z ^ (z >> 31)
    }
}

enum PropertyTests {
    static let alphabet: [Character] = ["a", "b", "c", "é"]

    static func generate(_ minLength: Int, _ maxLength: Int, _ rng: inout SeededRNG) -> String {
        let length = Int.random(in: minLength...maxLength, using: &rng)
        var result = ""
        for _ in 0..<length { result.append(alphabet.randomElement(using: &rng)!) }
        return result
    }

    static func oracle(_ a: String, _ b: String) -> Int {
        if a == b { return 0 }
        let left = Array(a.unicodeScalars)
        let right = Array(b.unicodeScalars)
        if left.isEmpty { return right.count }
        if right.isEmpty { return left.count }
        var previous = Array(0...right.count)
        for i in 1...left.count {
            var current = [i] + Array(repeating: 0, count: right.count)
            for j in 1...right.count {
                let cost = left[i - 1] == right[j - 1] ? 0 : 1
                current[j] = Swift.min(previous[j] + 1, current[j - 1] + 1, previous[j - 1] + cost)
            }
            previous = current
        }
        return previous[right.count]
    }

    static func runQuery(
        _ entries: [(String, UInt64?)], _ query: String, _ maximumDistance: Int
    ) throws -> [String: (Int, UInt64?)] {
        let dictionary = try DynamicDAWG()
        for (term, value) in entries { _ = try dictionary.put(term, value: value) }
        let transducer = try Transducer(dictionary: dictionary)
        var result: [String: (Int, UInt64?)] = [:]
        let cursor = try transducer.query(query, maximumDistance: maximumDistance)
        for match in cursor {
            guard case let .text(term) = match.term else { fatalError("expected text term") }
            result[term] = (match.distance, match.id)
        }
        transducer.close()
        dictionary.close()
        return result
    }

    static func run() throws {
        var rng = SeededRNG(seed: 20_260_809)

        let variants: [(dist: (String, String) -> Int, thr: (String, String, Int) -> Int?)] = [
            ({ EditDistance.levenshtein($0, $1) }, { EditDistance.levenshtein($0, $1, threshold: $2) }),
            ({ EditDistance.damerauOSA($0, $1) }, { EditDistance.damerauOSA($0, $1, threshold: $2) }),
            ({ EditDistance.damerauLevenshtein($0, $1) }, { EditDistance.damerauLevenshtein($0, $1, threshold: $2) }),
        ]

        // (a) distance self-consistency and oracle agreement.
        for _ in 0..<400 {
            let a = generate(0, 6, &rng)
            let b = generate(0, 6, &rng)
            let k = Int.random(in: 0...3, using: &rng)
            precondition(EditDistance.levenshtein(a, b) == oracle(a, b), "standard oracle agreement")
            for variant in variants {
                let full = variant.dist(a, b)
                precondition(variant.dist(b, a) == full, "symmetry")
                precondition(variant.dist(a, a) == 0, "identity")
                let bounded = variant.thr(a, b, k)
                precondition(bounded == (full <= k ? full : nil), "threshold consistency")
            }
        }

        // (b) query result set equals the oracle, with exact distances and (c)
        // value round-trips (including nil values).
        for _ in 0..<150 {
            var entries: [String: UInt64?] = [:]
            let count = Int.random(in: 0...8, using: &rng)
            for _ in 0..<count {
                let term = generate(0, 6, &rng)
                let value: UInt64? = Int.random(in: 0...3, using: &rng) == 0
                    ? nil : UInt64.random(in: 0...UInt64.max, using: &rng)
                entries.updateValue(value, forKey: term)
            }
            let query = generate(0, 6, &rng)
            let k = Int.random(in: 0...3, using: &rng)
            let got = try runQuery(entries.map { ($0.key, $0.value) }, query, k)

            var expected: [String: UInt64?] = [:]
            for (term, value) in entries where oracle(query, term) <= k {
                expected.updateValue(value, forKey: term)
            }
            precondition(Set(got.keys) == Set(expected.keys), "result set equals oracle")
            for (term, observed) in got {
                precondition(observed.0 == oracle(query, term), "exact distance")
                precondition(observed.1 == expected[term]!, "value round-trip")
            }
        }

        // (c) u64 boundaries pinned.
        for value in [UInt64(0), 1, UInt64(1) << 63, UInt64.max] {
            let got = try runQuery([("term", value)], "term", 0)
            let match = got["term"]!
            precondition(match.0 == 0 && match.1 == value, "u64 round-trip \(value)")
        }

        print("Swift property tests passed")
    }
}
